crates/bevy_ecs/src/system/commands/config.rs

```diff
@@ -0,0 +1,158 @@
+use std::{
+    fmt::Debug,
+    ops::{Deref, DerefMut},
+};
+
+use super::{fallible::*, Command};
+
+pub trait AddCommand {
+    fn add_command(&mut self, command: impl Command);
+}
+
+/// Provides configuration mechanisms in case a command errors.
+/// You can specify a custom handler via [`FallibleCommandConfig::on_failure`] or
+/// use one of the provided implementations.
+///
+/// ## Note
+/// The default error handler logs the error (via [`error!`]), but does not panic.
+pub struct FallibleCommandConfig<'a, C, T>
+where
+    C: FallibleCommand,
+    C::Error: Debug,
+    T: AddCommand,
+{
+    command: Option<C>,
+    inner: &'a mut T,
+}
+
+impl<'a, C, T> Deref for FallibleCommandConfig<'a, C, T>
+where
+    C: FallibleCommand,
+    C::Error: Debug,
+    T: AddCommand,
+{
+    type Target = T;
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        self.inner
+    }
+}
+
+impl<'a, C, T> DerefMut for FallibleCommandConfig<'a, C, T>
+where
+    C: FallibleCommand,
+    C::Error: Debug,
+    T: AddCommand,
+{
+    #[inline]
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.inner
+    }
+}
+
+/// Similar to [`FallibleCommandConfig`] however does not
+/// implement [`DerefMut`] nor return `&mut T` of the underlying
+/// Commands type.
+pub struct FinalFallibleCommandConfig<'a, C, T>
```

Comment:
Why would we want to use this, especially in contrast to the other trait?

Reply:
This doesn't implement the `DerefMut` trait so we can no longer chain commands.
Specifically is prevents this bug:
```rust
fn system(mut commands: Commands) {
    commands
        .entity(e)
        .despawn()
        .insert(42); // this is incorrect since the entity is despawned
}
```
`despawn()` returns the `FinalFallibleCommandConfig` so you can't keep chaining commands.